### PR TITLE
[WIP]Add NULS support

### PR DESCRIPTION
**Pull request by daviyang35**
- [x] Implement functionality in C++. Put it in a subfolder of `src/`.
    - [x] Address (if necessary)
    - [x] Transaction (if necessary)
    - [x] Signer
- [x] Write unit tests. Put them in a subfolder of `tests/`.
    [x] `Mnemonic phrase - > Address` derivation test. Put this test in the `CoinTests.cpp` file
- [x] Add relevant constants in `TWCoinType`, `TWP2SHPrefix`, `TWEthereymChainID`, `TWHRP`, etc., as necessary.
- [x] Return correct curve and purpose in `src/Coin.cpp`.
- [x] Implement address validation and derivation in `src/Coin.cpp`.
- [x] Implement coin configuration `src/include/TWCoinTypeConfiguration.cpp`.
- [x] Write interface header in `include/TrustWalletCore` and implement the interface in `src/interface`.
    - [x] Address interface (if necessary).
    - [x] Signing interface.
- [x] Validate generated code in Android an iOS projects. Write integration tests for each.

encode64LE(val & 0xffffffffffff, data) will write 8 byte data to buffer.
So we use encode48LE to write 6 byte number.

### CODE REVIEW

**Comment by hewigovens** [comment]
Thanks for the new pull request, could you please add a Kotlin or Swift transaction signing test case?

File: src/NULS/Address.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+// file LICENSE at the root of the source code distribution tree.
+#pragma once
+#include "../PrivateKey.h"
+#include "../PublicKey.h"
+
+#include <string>
+
+namespace TW::NULS {
+
+class Address {
```
Comment (hewigovens) [inline src/NULS/Address.h]: now we have a base `Base58Address`, could you please inherit `Address` from it? less code then

File: src/NULS/Signer.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+#include "../BinaryCoding.h"
+#include "../Hash.h"
+#include "../HexCoding.h"
+#include "../PrivateKey.h"
+
+using namespace TW;
+using namespace TW::NULS;
+
+/// Encodes a 48-bit little-endian value into the provided buffer.
+static inline void encode48LE(uint64_t val, std::vector<uint8_t>& data) {
```
Comment (hewigovens) [inline src/NULS/Signer.cpp]: let's move this to `src/NULS/BinaryCoding.h` like `src/Ripple/BinaryCoding.h`

File: src/NULS/TransactionBuilder.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        plan.set_to_address(purpose.to_address());
+        plan.set_remark(purpose.remark());
+
+        int64_t availableAmount = 0;
+        for (auto& utxo : purpose.utxos()) {
+            availableAmount += utxo.amount();
+        }
+        plan.set_available_amount(availableAmount);
+
+        // UTXO 输入 资金量小于转出金额，失败
```
Comment (hewigovens) [inline src/NULS/TransactionBuilder.h]: could you please use English? other people might not read Chinese comments

File: src/NULS/TransactionBuilder.h [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        Proto::TransactionOutput output;
+        output.set_amount(purpose.amount());
+        output.set_to_address(purpose.to_address());
+        output.set_index(0);
+        output.set_lock_time(0);
+
+        std::vector<Proto::TransactionOutput> outputs;
+        outputs.push_back(output);
+
+        // 对UTXO排序
```
Comment (hewigovens) [inline src/NULS/TransactionBuilder.h]: Same here and below

File: src/NULS/TransactionBuilder.h [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+#include "../proto/NULS.pb.h"
+
+#include <algorithm>
+
+namespace TW::NULS {
+
+struct TransactionBuilder {
+    /// Plans a transaction by selecting UTXOs and calculating fees.
+    static Proto::TransactionPlan plan(const NULS::Proto::TransactionPurpose& purpose) {
```
Comment (hewigovens) [inline src/NULS/TransactionBuilder.h]: Could you please add a test case for sending max amount?

**Comment by hewigovens** [comment]
Hi, can you fix CI?

File: src/NULS/TransactionBuilder.h (hunk 5 above)
Comment (daviyang35) [reply src/NULS/TransactionBuilder.h]: Nuls don't have sending max amount limit.
I add transaction max data size check.

Comment (hewigovens) [reply src/NULS/TransactionBuilder.h]: oh, consider this case: user wants to send all UTXOs, we have logic to adjust output count and fees

**Comment by hewigovens** [comment]
You might need to add a test case in `CoinType+Address.swift` to fix the iOS CI

**Comment by daviyang35** [comment]
@hewigovens max amount transfer plan support. check tests/NULS/TransactionPlanTest.cpp

**Pull request merged**